### DISCUSSION TITLE
Allow locked categories

### DISCUSSION THREAD

**Discussion by Andre601** [community/community#2937] [Discussions] [answered]
A nice feature would be to have "Locked categories"

These categories would essentially allow Users with Write/Push access to create a post in it. Would allow important announcements without the fear that people could fill the categorie with questions or topics not meant for it.

The current alternative would be to just post them in a general category and pin it which often isn't that great, especially if you already have pinned posts.

**Comment by becca** [comment]
We've heard this feedback quite a bit before, especially as it relates to having a category like `Announcements` where only select folks could post discussions. 

Are there any other use cases for customizing the granularity of read/write permissions at a category level that you can think of needing?

**Reply by dzervas** [reply]
A very strong connection would be with security advisories - be able to discuss in private and then optionally make the discussion public.

Also ability to "lock" discussions (e.g. after a discussion has turned into an editor war) but leave them readable by the public

**Reply by Mrlinkwii** [reply]
probably the wrong place for this but have an option like  "lock to contributors" options for general discussions , due to the fact the discussions may be used as a place for spam

**Reply by jarrodmillman** [reply]
In the scientific Python ecosystem, we use enhancement proposals (similar to Python's PEPs) in many projects (e.g., NumPy, scikit-image, NetworkX, etc.).  I've started looking into integrating the discussions in to this process and this feature would be very helpful.

The basic idea is anyone could submit an "Ideas" Discussion for a new enhancement proposal.  The new idea for a proposal would be discussed and the community could comment on it.  It the idea attracts interest and seems worth considering, the core developer team would move the discussion to the "Enhancement Proposal" category from the "Ideas" category.  At the same time, the idea would be given an enhancement proposal number (just like PEPs have).  The person who contributed the idea (or whoever else ends up volunteering to write the enhancement proposal) would be asked to create a PR with a draft of the new enhancement proposal using the assigned number and pointing to the "Enhancement Proposal" discussion.

It is annoying that anyone can submit a new discussion to the  "Enhancement Proposal" category currently.  Ideally, anyone to be able to discuss things in the  "Enhancement Proposal" category, but only the core developers/maintainers would be able to move items from the "Ideas" category to the "Enhancement Proposal" category.

**Reply by cassidyjames** [reply]
I think this falls under the category of announcements, but we're using a dedicated category called "Pinned" for our pinned Discussions, and it would be great to be able to limit that to specific people/teams/repo collaborators. https://github.com/elementary/edw/discussions/categories/pinned

**Reply by cassidyjames** [reply]
For context, we're using Discussions to plan an online conference. We could also use a locked category to move Discussions once they're accepted as a talk, which would be pretty slick: https://github.com/elementary/edw/discussions/categories/talk-proposals

**Comment by Andre601** [comment]
The only other thing would perhaps be a changelog category, but I would say that this is still quite close to announcements which doesn't make it that significantly different.

**Comment by StjerneIdioten** [comment]
Another thing I personally thought of was for a "development blog" category. I've adopted discussions as the "forum" of my project, so it would be great if I could have a locked category for development updates. It is of course similar to the "announcement" category, but I was thinking more for blog style posts similar to Paradox Development Diaries (if anyone plays eu4 😉) where as I use the "announcement" category solely as a place for changelogs or other big changes relating to the entire scope of the project.

**Reply by Andre601** [reply]
My idea is just about locking categories to people with push access.
There never was the idea of only a announcement category.

**Reply by StjerneIdioten** [reply]
I might have formulated myself badly then, because that wasn't the message I was trying to convey either. I was just trying to present another use case for locked (as in not open for everyone to post) categories other than the "announcement" category, which was mentioned in the first reply.

**Comment by noah-nuebling** [comment]
My use case would be posting Guides. 
I describe it in more detail [here](https://github.com/github/feedback/discussions/3692).

**Comment by b0o** [comment]
I'd like this feature as well!

I've got an idea to build a blog backed by Discussions, where discussions in a certain category become blog posts and responses become comments on the post. For this to work, there would need to be a distinction in the permissions between posting a new discussion and replying to an existing discussion.

**Comment by evi-liu** [comment]
## Announcement Category Format - Private Beta Available 🚀 

Thank you for all of the feedback on labels! I'm excited to announce that we have a private beta available, where you can create a new announcements category. In this category format, only admins and maintainers can post a new discussion to the category, but anyone can comment and reply. If you'd like to be included in the beta, please share the repository you'd like to be flagged in below. For now, this private beta is limited to the first 15 repositories that respond. 

![image](https://user-images.githubusercontent.com/8809849/116756934-e757b800-a9c1-11eb-874b-b36f7f534479.png)

**Reply by Andre601** [reply]
I hope outside contributors also count towards asking for access:
- https://github.com/botblock/JavaBotBloclAPI (Maintainer)
- https://github.com/PlaceholderAPI/PlaceholderAPI (Outside contributor)

**Reply by darbyjack** [reply]
> I hope outside contributors also count towards asking for access:
> 
> * [botblock/JavaBotBloclAPI](https://github.com/botblock/JavaBotBloclAPI) (Maintainer)
> * [PlaceholderAPI/PlaceholderAPI](https://github.com/PlaceholderAPI/PlaceholderAPI) (Outside contributor)

I am one of the maintainers for https://github.com/PlaceholderAPI/PlaceholderAPI and would love to help out testing this for you guys.

**Reply by b0o** [reply]
Awesome news! I would like to test it out on https://github.com/b0o/maddison.io

**Reply by cassidyjames** [reply]
We would love to try it at https://github.com/elementary/edw

**Reply by JRoy** [reply]
Would love to see this on [EssentialsX/Essentials](https://github.com/EssentialsX/Essentials)

**Reply by noah-nuebling** [reply]
Great news! :) I'd love to test this on 
https://github.com/noah-nuebling/mac-mouse-fix/

**Reply by evi-liu** [reply]
✅ all repos in this thread have been flagged in! so if you go to category customization, you should see the option for adding an announcements category now!

@Andre601 seems like botblock/JavaBotBloclAPI leads to a 404 repository. maybe there's a typo in the name you provided?

**Reply by Andre601** [reply]
> ✅ all repos in this thread have been flagged in! so if you go to category customization, you should see the option for adding an announcements category now!
> 
> @Andre601 seems like botblock/JavaBotBloclAPI leads to a 404 repository. maybe there's a typo in the name you provided?

@evi-liu You're right. It's botblock/JavaBotBlockAPI

**Reply by Andre601** [reply]
hey @evi-liu I encounter a weird issue with locked categories, which is that I can create, and edit categories to be locked, move existing discussions to them, but cannot create discussions in that category itself (Doesn't show in the category selection).

Is this a intentional behaviour of sorts? This happens on botblock/JavaBotBlockAPI
I'm Maintainer on the repo and team member on the org, but I'm not owner of either of them. This shouldn't matter tho as you mentioned that anyone with push access should be able to create such posts.
I can't tell if this behaviour is also the case in PlaceholderAPI/PlaceholderAPI because I believe the announcement category there hasn't been updated yet, but since I'm only an outside contributor can I most likely not create anything there anyways... I hope there will be a setting to set what access level can post stuff in such a category.

EDIT: One thing I also wonder is if the GitHub-Actions account together with the GitHub token is considered a Maintainer for discussions, as I consider preparing a Workflow that creates an automatic discussion [based on this style here](https://github.com/botblock/JavaBotBlockAPI/discussions/240) (compared to GitHub's option of creating a discussion with a release) whenever I make a new release.

**Reply by Andre601** [reply]
Could I also get https://github.com/Andre601/OneVersionRemake added for this?

**Comment by evi-liu** [comment, marked as answer]
## Announcements category available as public beta! 🚀 

Thanks for all of the helpful feedback and private beta testing of the announcements category. We've now rolled this out to all repositories with Discussions. Happy announcing and feel free to drop any feedback in this thread.

**Reply by Andre601** [reply]
Looks like the issue I mentioned before still hasn't been fixed.

Please forward #3912 to the responsible members of Discussions development, as not being able to create Discussions directly in a locked category as a Maintainer/Contributor is a general waste and only causes extra effort that really isn't needed.

Thanks.

**Reply by evi-liu** [reply]
Hi @Andre601, I'm not able to reproduce this bug. Are other maintainers in your repository also unable to create Discussions in the announcements category?

**Reply by Andre601** [reply]
I'm pretty much the only Maintainer in the JavaBotBlockAPI project, but @darbyjack has confirmed to me, that he too can't create posts in the locked category of the PlaceholderAPI repository.

Both repositories have been linked in #3912

**Reply by Andre601** [reply]
Something else to point out is, that I now enabled locked categories in [Andre601/OneVersionRemake](https://github.com/Andre601/OneVersionRemake) and there I seem to be able to create posts in locked categories while I still can't in the aforementioned repositories.

I have Maintainer status on JavaBotBlockAPI and are an outside contributor on PlaceholderAPI...
Could a possible reason be, that only the owner of a repository has the right to directly create posts there? Because neither I nor darbyjack are the owners of the respective Organisation the Repositories correspond to.

**Reply by wnbittle** [reply]
I'm seeing the same behavior in [dyn4j](https://github.com/dyn4j/dyn4j).  I'm the only maintainer.

![image](https://user-images.githubusercontent.com/11467997/118721525-faa7b780-b7f8-11eb-9ff1-ce02fad07adc.png)

I was not part of the closed beta.  I had a category called Announcements before and configured that category with the new Announcements setting.

**Reply by wnbittle** [reply]
Update - I realized I was not an Admin of the repo.  After changing to Admin, I can now create discussions in the locked category.  doh!

**Reply by Andre601** [reply]
I get the idea that this might be related to Organisations and whether you're the owner of it or not... I see no other reason why Maintainers don't have rights to post in those categories, but can create those and move posts to them...

**Reply by Andre601** [reply]
Also, for context: On JavaBotBlockAPI do I have the "java-libs" role in the organisation and that role has admin rights for the repository, so it can't be that for my case.
![image](https://user-images.githubusercontent.com/11576465/118722227-34de7c80-b82c-11eb-81d9-f2ed7ceafff3.png)

**Reply by wnbittle** [reply]
In my case, I have an Organization "dyn4j" where I'm the only member with a few repos under it.  In the [primary repo ](https://github.com/dyn4j/dyn4j/) my account wnbittle had only Write permissions.  After changing the permissions to Admin, the issue was resolved.  Not sure why the permissions were Write in the first place, they should have been Admin...

**Reply by Andre601** [reply]
Got the Owner of the botblock org @MattIPv4 to try and test if he can post in the locked category, which had the same result (He couldn't)

**Reply by MattIPv4** [reply]
👋 Repro'ed this in cdnjs/cdnjs -- I created a new category with this option and I am unable to post it in even whilst being an owner of the org and an admin on the repo:

![image](https://user-images.githubusercontent.com/12371363/118723215-0a3cf580-b825-11eb-8a0c-ead75b2cd330.png)

It looks like this is the result of incorrect permission calculations -- I am in the maintainers team and so have Admin access, and I am an org owner, but my account specifically was set to Write access in the repo, which appears to be what was preventing me from posting, though this would normally have no effect on other access as admin + owner takes precedence for all other actions.

![image](https://user-images.githubusercontent.com/12371363/118723777-b7177280-b825-11eb-95fc-904afa9c108d.png)

I set my account specifically to be an admin (as well as still being in a team that has admin + being an owner of the org), and now I am able to post in the category.

**Reply by MattIPv4** [reply]
@Andre601 only has access to the repo they are testing on through being in a team that has the Admin role in the repo -- my best guess with this bug is that permissions granted by teams aren't being respected at all, and this aligns with what I repro'ed in cdnjs/cdnjs.

I'm also the owner of that org as well as having the Admin role via a team on that repo, and as Andre noted, I was also unable to post in the category there. Definitely seems like posting is only checking individual account access on the repo, not considering teams/org ownership.

**Reply by evi-liu** [reply]
Ah yes, I think you're right that we only check for repository permissions and not team/org. I've passed this along to our engineering team, and we'll see what we can do to respect those permissions as well.

**Reply by ColinKerr** [reply]
We would like to be able to have a locked category that writers, maintainers and admins can use.  Today locked categories have a fixed set of roles, maybe the roles that are allowed to post a category could be configurable.

Our use case is that we want a category for developers to announce new features and ask for user feedback.  If we limit it to just maintainers and admins then developers would have to get their PM or one of the admins to make the post for them.  If we open the category to everyone then users may propose new features there.  

To avoid blocking our developers we made the category open and will just move feature requests to the idea category.

**Reply by laymonage** [reply]
^ In addition to the above, I'd also like to let installed GitHub Apps create a new discussion in the locked category. This would allow us to create automated announcements (and other possibilities).

**Reply by antn** [reply]
Hey folks! I've shipped a change that should fix the issue where you have permissions from a team, or from being an org owner. As part of this, GitHub Apps should also be able to create new announcement discussions as well.

Let me know if you're still running into any issues!

**Reply by laymonage** [reply]
@antn Thank you very much, it works on my GitHub App now! Also, could you take a look at #3968? It's not directly related but it would be awesome if it could be fixed. Thanks!

**Reply by IanMitchell** [reply]
> Hey folks! I've shipped a change that should fix the issue where you have permissions from a team, or from being an org owner. As part of this, GitHub Apps should also be able to create new announcement discussions as well.
> 
> Let me know if you're still running into any issues!

Hi @antn - Discord just turned on locked announcement categories in our repo, and non-admin/maintainer users are reporting that they're able to create a discussion under an open category, then edit the discussion and select the locked category successfully. Is this a misconfiguration on our part, or maybe a bug?

**Reply by MattIPv4** [reply]
Can confirm this (^) is reproducible:

The `News` category in my test repo repo is locked:

![image](https://user-images.githubusercontent.com/12371363/124502301-20921700-ddbb-11eb-80ea-de99f1d903e1.png)

When posting as a regular user, you cannot post to it:

![image](https://user-images.githubusercontent.com/12371363/124502156-d741c780-ddba-11eb-884a-40ca4f5ffd2a.png)

Once posted though, as a regular user, you can open the category menu and choose the locked `News` category:

![image](https://user-images.githubusercontent.com/12371363/124502172-df016c00-ddba-11eb-87f9-42b9853258aa.png)
![image](https://user-images.githubusercontent.com/12371363/124502184-e294f300-ddba-11eb-832b-fac2bc1de562.png)

This does indeed successfully move it to the locked category, as a regular user:

![image](https://user-images.githubusercontent.com/12371363/124502203-ecb6f180-ddba-11eb-9559-bada371e88db.png)

**Reply by antn** [reply]
Sorry about that, @IanMitchell and @MattIPv4! I just deployed a fix, so that shouldn't be possible anymore. This appears to have been a bug with the initial implementation of the announcements feature, and not the recent changes – but great catch! 😱

**Reply by IanMitchell** [reply]
Thanks so much! Appreciate the quick fix!

**Comment by ghost** [comment]
I just pray it's dose not happen again
